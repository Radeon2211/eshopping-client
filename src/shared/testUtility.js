// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const historyPageNum = (pageNumber) => ({
  listen: jest.fn(),
  location: { search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push: jest.fn(),
});

export const propsPagination = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  maxQuantityPerPage: 2,
});
