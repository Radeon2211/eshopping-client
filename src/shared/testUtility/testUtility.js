// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';
import { v4 as uuidv4 } from 'uuid';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../../store/reducers/rootReducer';
import { productConditions } from '../constants';

export const defaultDeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: 'active',
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const createHistoryPageNumber = (pageNumber = 1, push = jest.fn()) => ({
  listen: jest.fn(),
  location: { pathname: '/products', search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push,
});

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
  quantitySold = 0,
  buyerQuantity = 0,
  description = '',
  condition = productConditions.NEW,
) => ({
  _id: productId || uuidv4(),
  name,
  price,
  quantity,
  quantitySold,
  buyerQuantity,
  description,
  condition,
  photo,
  seller: {
    username: sellerUsername,
  },
  __v: 0,
  createdAt: '2021-02-10T19:10:38.872Z',
  updatedAt: '2021-02-10T19:10:38.872Z',
});

export const createCartItem = (
  sellerUsername = 'username',
  quantity = 1,
  productId,
  price = 2,
  name = 'product name',
  productQuantity = 2,
  photo = false,
) => {
  const product = createProductItem(
    productId || uuidv4(),
    sellerUsername,
    productQuantity,
    price,
    name,
    photo,
  );

  return {
    _id: uuidv4(),
    quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
) => ({
  _id: productId || uuidv4(),
  name,
  photo,
  price,
  quantity,
  seller: {
    username: sellerUsername,
  },
});

export const createOrder = (
  products = [createTransactionAndOrderProdItem()],
  orderId = uuidv4(),
  sellerUsername,
  buyerUsername,
  overallPrice,
  createdAt,
  sellerEmail,
  sellerPhone,
) => {
  const seller = sellerUsername
    ? {
        username: sellerUsername,
        email: sellerEmail,
        phone: sellerPhone,
      }
    : null;

  const buyer = buyerUsername
    ? {
        username: buyerUsername,
      }
    : null;

  return {
    _id: orderId,
    seller,
    buyer,
    products,
    overallPrice,
    deliveryAddress: defaultDeliveryAddress,
    createdAt,
    updatedAt: createdAt,
    __v: 0,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...defaultState.product,
      ...initialProduct,
    },
    ui: {
      ...defaultState.ui,
      ...initialUI,
    },
  };
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

export const checkReqMethodAndURL = (moxiosInstance, method, url, reqNumber = 0) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests[reqNumber - 1];
    return method === prevRequest.config.method && url === prevRequest.config.url;
  }
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState,
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};
