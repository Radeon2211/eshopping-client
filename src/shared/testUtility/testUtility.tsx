/* eslint-disable import/no-extraneous-dependencies */
import { ReactElement } from 'react';
import { Router } from 'react-router-dom';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import { MockStoreEnhanced } from 'redux-mock-store';
import { fireEvent, waitFor, render } from '@testing-library/react';
import { v4 as uuidv4 } from 'uuid';
import moxios from 'moxios';
import rootReducer from '../../store/reducers/rootReducer';
import theme from '../../styled/theme';
import {
  CartItem,
  DeliveryAddress,
  Order,
  OrderProductItem,
  Product,
  Profile,
  ProfileStatus,
} from '../types/types';
import { ProductCondition } from '../types/enums';

export const defaultDeliveryAddress: DeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile: Profile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: ProfileStatus.ACTIVE,
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (props: Partial<Product> = {}): Product => {
  return {
    _id: uuidv4(),
    quantity: 1,
    price: 2,
    name: 'product name',
    photo: false,
    quantitySold: 0,
    buyerQuantity: 0,
    description: '',
    condition: ProductCondition.NEW,
    seller: {
      username: 'username',
    },
    createdAt: '2021-02-10T19:10:38.872Z',
    updatedAt: '2021-02-10T19:10:38.872Z',
    __v: 0,
    ...props,
  };
};

export const createCartItem = (props = {}): CartItem => {
  const finalProps = {
    id: uuidv4(),
    sellerUsername: 'username',
    quantity: 1,
    productId: uuidv4(),
    price: 2,
    name: 'product name',
    productQuantity: 2,
    photo: false,
    ...props,
  };
  const product = createProductItem({
    _id: finalProps.productId,
    quantity: finalProps.productQuantity,
    price: finalProps.price,
    name: finalProps.name,
    photo: finalProps.photo,
    seller: {
      username: finalProps.sellerUsername,
    },
  });

  return {
    _id: finalProps.id,
    quantity: finalProps.quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  props: Partial<OrderProductItem> = {},
): OrderProductItem => {
  return {
    _id: uuidv4(),
    quantity: 1,
    price: 2,
    name: 'product name',
    photo: false,
    seller: {
      username: 'username',
    },
    ...props,
  };
};

export const createOrder = (props: Partial<Order> = {}): Order => {
  return {
    _id: uuidv4(),
    overallPrice: 100,
    seller: null,
    buyer: null,
    products: [createTransactionAndOrderProdItem()],
    deliveryAddress: defaultDeliveryAddress,
    createdAt: '2021-02-10T19:10:38.872Z',
    updatedAt: '2021-02-10T19:10:38.872Z',
    __v: 0,
    ...props,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...(defaultState.product as object),
      ...initialProduct,
    },
    ui: {
      ...(defaultState.ui as object),
      ...initialUI,
    },
  };
  // @ts-expect-error TBF after redux upgrade
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

type MoxiosRequestsItems = { config: { method: string; url: string } }[];
type MockedMoxiosInstance = {
  requests: {
    __items?: MoxiosRequestsItems;
    mostRecent?: () => { config: { method: string; url: string } };
  };
};
type DefaultMoxiosInstance = Omit<typeof moxios, 'requests'> & {
  requests: typeof moxios.requests & {
    __items?: MoxiosRequestsItems;
  };
};
export const checkReqMethodAndURL = (
  moxiosInstance: MockedMoxiosInstance | DefaultMoxiosInstance,
  method: string,
  url: string,
  reqNumber = 0,
) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests?.[reqNumber - 1];
    return prevRequest && method === prevRequest.config.method && url === prevRequest.config.url;
  }
  if (!moxiosInstance.requests.mostRecent) return false;
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState: {
    auth?: object;
    product?: object;
    ui?: object;
  },
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};

export const clickAtSubmitButton = async (container: Element) => {
  await waitFor(() => {
    const submitBtn = container.querySelector('button[type="submit"]');
    if (submitBtn) {
      fireEvent.click(submitBtn);
    }
  });
};

export const testRouterPushCall = (
  pushFn: ReturnType<typeof jest.fn>,
  callIdx: number,
  expectedPathname: string,
  expectedSearch = '',
) => {
  expect(pushFn.mock.calls[callIdx][0].pathname).toEqual(expectedPathname);
  expect(pushFn.mock.calls[callIdx][0].search).toEqual(expectedSearch);
};

type RenderAppPartOptions = {
  push?: ReturnType<typeof jest.fn>;
  pathname?: string;
  search?: string;
  withoutThemeProvider?: boolean;
  withoutRouter?: boolean;
  store?: MockStoreEnhanced<object, object>;
  onlyReturnWrappedElement?: boolean;
};

export const renderAppPart = (element: ReactElement, options: RenderAppPartOptions = {}) => {
  const navigator = {
    createHref: jest.fn(),
    push: options.push || jest.fn(),
    go: jest.fn(),
    replace: jest.fn(),
  };

  const location = {
    pathname: options.pathname || '/',
    search: options.search || '',
  };

  let content = element;
  if (!options.withoutThemeProvider) {
    content = <ThemeProvider theme={theme}>{element}</ThemeProvider>;
  }
  if (options.store) {
    content = <Provider store={options.store}>{content}</Provider>;
  }
  if (!options.withoutRouter) {
    content = (
      <Router location={location} navigator={navigator}>
        {content}
      </Router>
    );
  }

  return options.onlyReturnWrappedElement ? content : render(content);
};
