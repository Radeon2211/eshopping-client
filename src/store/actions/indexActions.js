export {
  registerUser,
  loginUser,
  fetchProfile,
  logoutUser,
  fetchOtherUser,
  setOtherUser,
  updateUser,
  changeDeliveryAddress,
  deleteAccount,
  addAdmin,
  removeAdmin,
  sendAccountVerificationLink,
  resetPassword,
} from './authActions';

export {
  fetchCart,
  addCartItem,
  updateCartItem,
  clearCart,
  removeCartItem,
  setTransaction,
  goToTransaction,
  buyProducts,
} from './tradeActions';

export { setModal, deleteMessage, changeProductsPerPage } from './uiActions';

export {
  addProduct,
  editProduct,
  fetchProducts,
  fetchProductDetails,
  setProductDetails,
  deleteProduct,
} from './productActions';

export { fetchOrders, fetchOrderDetails, setOrderDetails } from './orderActions';
