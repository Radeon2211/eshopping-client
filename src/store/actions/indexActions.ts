export {
  registerUser,
  loginUser,
  fetchProfile,
  logoutUser,
  fetchOtherUser,
  setOtherUser,
  updateUser,
  changeEmail,
  changeDeliveryAddress,
  deleteAccount,
  addAdmin,
  removeAdmin,
  sendAccountVerificationLink,
  resetPassword,
} from './authActions/authActions';

export {
  fetchCart,
  addCartItem,
  updateCartItem,
  clearCart,
  removeCartItem,
  setTransaction,
  goToTransaction,
  buyProducts,
} from './tradeActions/tradeActions';

export { setMessage, setModal, changeProductsPerPage } from './uiActions/uiActions';

export {
  addProduct,
  editProduct,
  fetchProducts,
  fetchProductDetails,
  setProductDetails,
  deleteProduct,
} from './productActions/productActions';

export { fetchOrders, fetchOrderDetails, setOrderDetails } from './orderActions/orderActions';
