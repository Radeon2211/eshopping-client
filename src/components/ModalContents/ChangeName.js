import React, { useCallback } from 'react';
import { Formik } from 'formik';
import { useDispatch, useSelector } from 'react-redux';
import * as actions from '../../store/actions/indexActions';
import Form from '../UI/Form/Form';
import Input from '../UI/Input/Input';
import { inputKinds } from '../../shared/constants';

const ChangeName = () => {
  const userProfile = useSelector((state) => state.auth.profile);

  const dispatch = useDispatch();
  const onChangeName = useCallback((creds) => dispatch(actions.changeName(creds)), [dispatch]);

  return (
    <Formik
      initialValues={{
        firstName: userProfile?.firstName,
        lastName: userProfile?.lastName,
      }}
      onSubmit={(data) => {
        onChangeName(data);
      }}
    >
      {({ dirty }) => (
        <Form btnText="change" headingText="Change your name" isValid={dirty} cancellable>
          <Input
            kind={inputKinds.INPUT}
            config={{
              type: 'text',
              name: 'firstName',
              id: 'firstName',
              placeholder: 'Your first name',
              autoComplete: 'given-name',
            }}
            label="First name"
          />
          <Input
            kind={inputKinds.INPUT}
            config={{
              type: 'text',
              name: 'lastName',
              id: 'lastName',
              placeholder: 'Your last name',
              autoComplete: 'family-name',
            }}
            label="Last name"
          />
        </Form>
      )}
    </Formik>
  );
};

export default ChangeName;
