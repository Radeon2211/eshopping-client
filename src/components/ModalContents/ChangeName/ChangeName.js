import React, { useCallback } from 'react';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { useDispatch, useSelector } from 'react-redux';
import * as actions from '../../../store/actions/indexActions';
import Form from '../../UI/Form/Form';
import Input from '../../UI/Input/Input';
import { inputKinds, userRules } from '../../../shared/constants';
import { getChangedValues } from '../../../shared/utility/utility';

const validationSchema = Yup.object({
  firstName: userRules.firstName,
  lastName: userRules.lastName,
});

export default function ChangeName() {
  const userProfile = useSelector((state) => state.auth.profile);

  const dispatch = useDispatch();
  const onUpdateUser = useCallback(
    (creds, message) => dispatch(actions.updateUser(creds, message)),
    [dispatch],
  );

  const initialValues = {
    firstName: userProfile.firstName,
    lastName: userProfile.lastName,
  };

  return (
    <Formik
      initialValues={initialValues}
      validationSchema={validationSchema}
      onSubmit={(data) => {
        const changedValues = getChangedValues(data, initialValues);
        onUpdateUser(changedValues, 'Name has been changed successfully');
      }}
    >
      {({ dirty, errors, touched, isValid, setFieldTouched }) => (
        <Form
          btnText="change"
          headingText="Change your name"
          isValid={dirty && isValid}
          cancellable
        >
          <Input
            kind={inputKinds.INPUT}
            config={{
              type: 'text',
              name: 'firstName',
              id: 'firstName',
              placeholder: 'Your first name',
              autoComplete: 'given-name',
              autoFocus: true,
              onInput: setFieldTouched.bind(this, 'firstName', true, true),
              'data-testid': 'ChangeName-firstName',
            }}
            label="First name"
            isValid={!errors.firstName}
            isTouched={touched.firstName}
          />
          <Input
            kind={inputKinds.INPUT}
            config={{
              type: 'text',
              name: 'lastName',
              id: 'lastName',
              placeholder: 'Your last name',
              autoComplete: 'family-name',
              onInput: setFieldTouched.bind(this, 'lastName', true, true),
              'data-testid': 'ChangeName-lastName',
            }}
            label="Last name"
            isValid={!errors.lastName}
            isTouched={touched.lastName}
          />
        </Form>
      )}
    </Formik>
  );
}
