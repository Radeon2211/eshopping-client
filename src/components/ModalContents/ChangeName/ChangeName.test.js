import React from 'react';
import { render, cleanup, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ChangeName from './ChangeName';
import theme from '../../../styled/theme';
import { clickAtSubmitButton, defaultUserProfile } from '../../../shared/testUtility/testUtility';
import * as actions from '../../../store/actions/indexActions';

const mockStore = configureMockStore([thunk]);

const oldFirstName = 'oldFirstName';
const oldLastName = 'oldLastName';

const setUp = () => {
  const store = mockStore({
    auth: {
      profile: { ...defaultUserProfile, firstName: oldFirstName, lastName: oldLastName },
    },
    ui: {
      isFormLoading: false,
      formError: '',
    },
  });
  store.dispatch = jest.fn();

  return {
    ...render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <ChangeName />
        </ThemeProvider>
      </Provider>,
    ),
    store,
  };
};

jest.mock('../../../store/actions/indexActions.js', () => ({
  ...jest.requireActual('../../../store/actions/indexActions.js'),
  updateUser: (credentials, message) => ({
    credentials,
    message,
  }),
}));

afterEach(cleanup);

describe('<ChangeName />', () => {
  describe('check how renders', () => {
    it('should render everything correctly', () => {
      const { asFragment } = setUp();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('check form', () => {
    const defaultNewFirstName = 'newFirstName';
    const defaultNewLastName = 'newLastName';

    it('should have values from state in inputs and focus on firstName input by default', () => {
      setUp();
      const firstNameInput = screen.getByTestId('ChangeName-firstName');
      const lastNameInput = screen.getByTestId('ChangeName-lastName');
      expect(firstNameInput).toHaveFocus();
      expect(firstNameInput.value).toEqual(oldFirstName);
      expect(lastNameInput.value).toEqual(oldLastName);
    });

    describe('should call', () => {
      const updateMessage = 'Name has been changed successfully';

      it('should call updateUser() with all changed values after inputs submit and button click', async () => {
        const { store, container } = setUp();

        const firstNameInput = screen.getByTestId('ChangeName-firstName');
        const lastNameInput = screen.getByTestId('ChangeName-lastName');

        const dataToPass = {
          firstName: defaultNewFirstName,
          lastName: defaultNewLastName,
        };

        await waitFor(() => {
          fireEvent.change(firstNameInput, { target: { value: defaultNewFirstName } });
        });
        await waitFor(() => {
          fireEvent.change(lastNameInput, { target: { value: defaultNewLastName } });
        });
        expect(firstNameInput.value).toEqual(defaultNewFirstName);
        expect(lastNameInput.value).toEqual(defaultNewLastName);

        await clickAtSubmitButton(container);
        expect(store.dispatch).toHaveBeenNthCalledWith(
          1,
          actions.updateUser(dataToPass, updateMessage),
        );

        await waitFor(() => {
          fireEvent.submit(firstNameInput);
        });
        expect(store.dispatch).toHaveBeenNthCalledWith(
          2,
          actions.updateUser(dataToPass, updateMessage),
        );

        await waitFor(() => {
          fireEvent.submit(lastNameInput);
        });
        expect(store.dispatch).toHaveBeenNthCalledWith(
          3,
          actions.updateUser(dataToPass, updateMessage),
        );

        expect(store.dispatch).toHaveBeenCalledTimes(3);
      });

      it('should call updateUser() with updated firstName only', async () => {
        const { store, container } = setUp();

        const firstNameInput = screen.getByTestId('ChangeName-firstName');

        const dataToPass = {
          firstName: defaultNewFirstName,
        };

        await waitFor(() => {
          fireEvent.change(screen.getByTestId('ChangeName-firstName'), {
            target: { value: defaultNewFirstName },
          });
        });
        expect(firstNameInput.value).toEqual(defaultNewFirstName);

        await clickAtSubmitButton(container);

        expect(store.dispatch).toHaveBeenCalledWith(actions.updateUser(dataToPass, updateMessage));
      });

      it('should call updateUser() with updated lastName only', async () => {
        const { store, container } = setUp();

        const lastNameInput = screen.getByTestId('ChangeName-lastName');

        const dataToPass = {
          lastName: defaultNewLastName,
        };

        await waitFor(() => {
          fireEvent.change(lastNameInput, {
            target: { value: defaultNewLastName },
          });
        });
        expect(lastNameInput.value).toEqual(defaultNewLastName);

        await clickAtSubmitButton(container);

        expect(store.dispatch).toHaveBeenCalledWith(actions.updateUser(dataToPass, updateMessage));
      });
    });

    describe('should NOT call', () => {
      it('should NOT call updateUser() if both inputs are empty', async () => {
        const { store, container } = setUp();

        const firstNameInput = screen.getByTestId('ChangeName-firstName');
        const lastNameInput = screen.getByTestId('ChangeName-lastName');

        await waitFor(() => {
          fireEvent.change(firstNameInput, {
            target: { value: '' },
          });
        });
        await waitFor(() => {
          fireEvent.change(lastNameInput, {
            target: { value: '' },
          });
        });
        expect(firstNameInput.value).toEqual('');
        expect(lastNameInput.value).toEqual('');

        await clickAtSubmitButton(container);

        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('should NOT call updateUser() if both values are the same as current', async () => {
        const { store, container } = setUp();
        await clickAtSubmitButton(container);
        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('should NOT call updateUser() if firstName input is empty (lastName is changed)', async () => {
        const { store, container } = setUp();

        const firstNameInput = screen.getByTestId('ChangeName-firstName');
        const lastNameInput = screen.getByTestId('ChangeName-lastName');

        await waitFor(() => {
          fireEvent.change(firstNameInput, {
            target: { value: '' },
          });
        });
        await waitFor(() => {
          fireEvent.change(lastNameInput, {
            target: { value: defaultNewLastName },
          });
        });
        expect(firstNameInput.value).toEqual('');
        expect(lastNameInput.value).toEqual(defaultNewLastName);

        await clickAtSubmitButton(container);

        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('should NOT call updateUser() if lastName input is empty (firstName is changed)', async () => {
        const { store, container } = setUp();

        const firstNameInput = screen.getByTestId('ChangeName-firstName');
        const lastNameInput = screen.getByTestId('ChangeName-lastName');

        await waitFor(() => {
          fireEvent.change(screen.getByTestId('ChangeName-firstName'), {
            target: { value: defaultNewFirstName },
          });
        });
        await waitFor(() => {
          fireEvent.change(screen.getByTestId('ChangeName-lastName'), {
            target: { value: '' },
          });
        });
        expect(firstNameInput.value).toEqual(defaultNewFirstName);
        expect(lastNameInput.value).toEqual('');

        await clickAtSubmitButton(container);

        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('should NOT call updateUser() if firstName is too long', async () => {
        const { store, container } = setUp();

        const firstNameInput = screen.getByTestId('ChangeName-firstName');

        let newFirstName = '';
        for (let i = 1; i <= 61; i += 1) {
          newFirstName += 'e';
        }
        await waitFor(() => {
          fireEvent.change(firstNameInput, {
            target: { value: newFirstName },
          });
        });
        expect(firstNameInput.value).toEqual(newFirstName);

        await clickAtSubmitButton(container);

        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('should NOT call updateUser() if lastName is too long', async () => {
        const { store, container } = setUp();

        const lastNameInput = screen.getByTestId('ChangeName-lastName');

        let newLastName = '';
        for (let i = 1; i <= 81; i += 1) {
          newLastName += 'e';
        }
        await waitFor(() => {
          fireEvent.change(lastNameInput, {
            target: { value: newLastName },
          });
        });
        expect(lastNameInput.value).toEqual(newLastName);

        await clickAtSubmitButton(container);

        expect(store.dispatch).not.toHaveBeenCalled();
      });
    });
  });
});
