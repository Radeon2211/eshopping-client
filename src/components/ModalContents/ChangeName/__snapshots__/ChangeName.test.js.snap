// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangeName /> check how renders should render everything correctly 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  justify-self: center;
  max-width: 100%;
  text-align: left;
  width: 65rem;
}

.c10 {
  box-shadow: 0 0 3px rgba(0,0,0,0.03);
  border-radius: 1px;
  cursor: pointer;
  font-weight: 700;
  line-height: 1.2;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  outline: none;
  position: relative;
  text-align: center;
  text-transform: uppercase;
  -webkit-transition: all 0.1s;
  transition: all 0.1s;
  font-size: 1.4rem;
  padding: 0.8rem 1.2rem;
  background-color: transparent;
  border: 2px solid #1E8BC3;
  color: #1E8BC3;
}

.c10:hover {
  border-color: #22A7F0;
  color: #22A7F0;
}

.c11 {
  box-shadow: 0 0 3px rgba(0,0,0,0.03);
  border-radius: 1px;
  cursor: pointer;
  font-weight: 700;
  line-height: 1.2;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  outline: none;
  position: relative;
  text-align: center;
  text-transform: uppercase;
  -webkit-transition: all 0.1s;
  transition: all 0.1s;
  font-size: 1.4rem;
  padding: 0.8rem 1.2rem;
  background-color: #1E8BC3;
  border: 2px solid #1E8BC3;
  color: #fff;
  background-color: #67809F;
  border-color: #67809F;
  color: #FAFAFA;
  cursor: not-allowed;
}

.c11:hover {
  background-color: #22A7F0;
  border-color: #22A7F0;
}

.c11:hover {
  background-color: #67809F;
  border-color: #67809F;
  color: #FAFAFA;
}

.c9 {
  display: block;
  font-style: normal;
  font-weight: normal;
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  margin-bottom: 0;
  margin-left: 0;
  margin-right: 2.4rem;
  overflow: visible;
  text-overflow: clip;
  white-space: normal;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  width: 100%;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c8 > *:not(:last-child) {
  margin-right: 2.4rem;
}

.c1 {
  font-size: 2.4rem;
  font-family: Noto Sans,sans-serif;
  line-height: 1;
  margin-bottom: 2.4rem;
  margin-top: 0;
  text-align: left;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 2.4rem;
}

.c3 .select {
  font-size: 1.4rem;
}

.c3 .textarea {
  border: none;
  border-bottom: 3px solid rgba(0,0,0,0.1);
  outline: none;
  font-size: 1.4rem;
  font-family: inherit;
  padding: 0 0 0.8rem 0;
  -webkit-transition: border-bottom 0.1s;
  transition: border-bottom 0.1s;
  resize: none;
}

.c3 .textarea::-webkit-input-placeholder {
  color: #BFBFBF;
}

.c3 .textarea:focus {
  border-bottom: 3px solid rgba(0,0,0,0.35);
}

.c3.valid .textarea {
  border-bottom: 3px solid #009680;
}

.c3.invalid .textarea {
  border-bottom: 3px solid #F22613;
}

.c7 {
  border: none;
  border-bottom: 3px solid rgba(0,0,0,0.1);
  color: #000;
  font-size: 1.3rem;
  font-family: inherit;
  outline: none;
  padding: 0 0 0.8rem 0;
  -webkit-transition: border-bottom 0.1s;
  transition: border-bottom 0.1s;
  width: 100%;
}

.c7::-webkit-input-placeholder {
  color: #BFBFBF;
}

.c7:focus {
  border-bottom: 3px solid rgba(0,0,0,0.35);
}

.c2.valid .c6 {
  border-bottom: 3px solid #009680;
}

.c2.invalid .c6 {
  border-bottom: 3px solid #F22613;
}

.c5 {
  font-size: 1.5rem;
  padding-bottom: 0.8rem;
  -webkit-transition: color 0.1s;
  transition: color 0.1s;
}

.c2.valid .c4 {
  color: #009680;
}

.c2.invalid .c4 {
  color: #F22613;
}

<div
    class="c0"
  >
    <h3
      class="c1"
      font-size=""
    >
      Change your name
    </h3>
    <form
      action="#"
      class="formik-form"
    >
      <div
        class="c2 c3"
        type="text"
      >
        <label
          class="c4 c5"
          for="firstName"
        >
          First name
        </label>
        <input
          autocomplete="given-name"
          class="c6 c7"
          data-testid="ChangeName-firstName"
          id="firstName"
          name="firstName"
          placeholder="Your first name"
          type="text"
          value="oldFirstName"
        />
      </div>
      <div
        class="c2 c3"
        type="text"
      >
        <label
          class="c4 c5"
          for="lastName"
        >
          Last name
        </label>
        <input
          autocomplete="family-name"
          class="c6 c7"
          data-testid="ChangeName-lastName"
          id="lastName"
          name="lastName"
          placeholder="Your last name"
          type="text"
          value="oldLastName"
        />
      </div>
      <div
        class="c8"
        direction="row"
        spacing="3"
        width="100%"
        wrap="nowrap"
      >
        <span
          class="c9"
          color=""
          display="block"
          overflow="visible"
          spacing="0px"
          transform=""
        >
          <button
            class="c10"
            color="blue"
            type="button"
          >
            cancel
          </button>
        </span>
        <button
          class="c11"
          color="blue"
          data-testid="Form-submit-btn"
          disabled=""
          type="submit"
        >
          change
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
